### ISSUE TITLE
Errors Cannot redeclare composerRequire when running commit on OSX  (+workaround)

### ISSUE THREAD

**Issue by jflaflamme** [phpro/grumphp#377]
| Q               | A
| --------------- | ---
| Version         | 0.11.6
| Bug?            | yes
| New feature?    | no
| Question?       | no
| Documentation?  | no

**Steps to reproduce:**
```git commit```

**Result:**
```PHP Fatal error:  Cannot redeclare composerRequire15ae15a0256f312901fce873ae77b254() (previously declared in /Users/***/Documents/Projects/***/vendor/composer/autoload_real.php:63) in /Users/***/Documents/Projects/***/vendor/composer/autoload_real.php on line 70```

----
FIX?
Tried, composer update, removing vendor, dumpautoload, never worked .. but when I removed **exec** from the git hooks everything was working as it should. Probably OSX exec related.

Hope this helps.

**Comment by veewee** [comment]
Hi @jflaflamme,

I doubt that it is OSX exec related. It looks like the  autoload_real.php file is loaded twice.

- Can you follow the autoload links in https://github.com/phpro/grumphp/blob/master/bin/grumphp#L6-L11 and make sure that no file is loaded twice?
- Can you give me some additional information about your setup?
- Did you install grumphp both in the project and globally? (You might want to remove the global one in that case.)
- What does your failing git hook look like?
- What PHP version are you on?


Thanks!

**Comment by jflaflamme** [comment]
Hello @veewee 

Only one is in the autoload links and I use only in a project mode with php 7.0.19.
Again, I can run it without exec, with the same setup and php.

`(cd "./" && printf "%s\n" "${DIFF}" | vendor/phpro/grumphp/bin/grumphp git:pre-commit --skip-success-output)` 

This works fine!  But if I put back with the exec as shipped, I got the error mentioned. 
I simply figured this out because running `vendor/phpro/grumphp/bin/grumphp git:pre-commit --skip-success-output` works fine, only the git hook wasn't.

**doesn't work (cannot redeclare error with composer require)**

`(cd "./" && printf "%s\n" "${DIFF}" | exec 'vendor/phpro/grumphp/bin/grumphp' 'git:commit-msg' "--git-user=$GIT_USER" "--git-email=$GIT_EMAIL" "$COMMIT_MSG_FILE")`

**work**
`(cd "./" && printf "%s\n" "${DIFF}" | vendor/phpro/grumphp/bin/grumphp git:commit-msg "--git-user=$GIT_USER" "--git-email=$GIT_EMAIL" "$COMMIT_MSG_FILE")`

Hope this helps.

**Comment by veewee** [comment]
There's nothing strange on your setup. I am also working on mac and never had the issue you described. We will need a way to reproduce this in order to fix the issue. The exec command was added a long time ago to be compatible with some Windows systems IIRC, so it's not something we could let out of the hook. 

A small workaround for your local config would be to overwrite the hooks_dir parameter as described in https://github.com/phpro/grumphp/blob/master/doc/parameters.md

**Comment by jflaflamme** [comment]
I understand, but right now I have no issues since I modified in .git/hooks/
That is not something that will be committed or changed unless I issue a re-init of grumphp

What would be the advantage of modifying the parameters ? I am not sure to follow

**Comment by veewee** [comment]
Since the path system got a rewrite in #644 , I am closing this issue for now.
Feel free to test that PR out and validate if it still works on your environment.
A new release will follow soon-ish.


If you still encounter this issue : try removing your vendor directory+reinstall it. The reason why this error occurs is because you load 2 separate autoload files with the same autoloader suffix. If you remove them, composer generates a new suffix.

**Comment by jerowork** [comment]
@veewee @jflaflamme FYI:

I do have this problem too and found out an interesting thing (am working on MacOS):

The redeclare error is happening (at my side) because the generated function is indeed declared multiple times. This is because the `autoload_real.php` is loaded twice. 

At first sight, this is strange because `bin/grumphp` is using `require_once`, which should not require an already loaded file. 

But when you are working on a **case sensitive** volume, `getcwd` can return something different than `__DIR__`.

Difference on my system:

- `getcwd`: `/Volumes/Devp/.../.../vendor/...`
- `__DIR__`: `/Volumes/DevP/.../.../vendor/...`

Also weird, the first char of the volume name is still in uppercase, but the last isn't. 

Did not find out yet why there is a difference between the two. Still figuring out.

@jflaflamme Do you work on a case-sensitive volume?

I am not sure what the solution should be, because I don't know exactly why the `bin/grumphp` is using both `getcwd` and `__DIR__`.

**Comment by jflaflamme** [comment]
@jerowork yes I do. Sorry for the delay.

**Comment by andycowan** [comment]
I hit this issue also - @jerowork is right, there is a difference between the results of ```cwd``` and ```___DIR___``` on a case-sensitive drive.

I was able to workaround the issue by changing the cwd() on line 11 in grumphp/bin/grumphp to a __DIR__

```
mother git:(feat-refactor-dependencies) ✗ git commit -m 'fix: refactor paysys/eurotest to remove dep' .
PHP Fatal error:  Cannot redeclare composerRequireefa666f063c78e72c6e4fe10321f938f() (previously declared in /Users/andrew/PS/Code/mother/vendor/composer/autoload_real.php:68) in /Users/andrew/PS/code/mother/vendor/composer/autoload_real.php on line 68
```

Note: Code and code in the path in the error message

**Comment by veewee** [comment]
cwd doesnt have to be the same as __DIR__ though.
It looks like a PHP bug and not something we would know inside grumphp...  Any thoughts why it could differ?

**Comment by jerowork** [comment]
Some time ago already since I had noticed this :).

In the meanwhile I just changed the volume name by removing the capitals, as I found out that it only occurs on volume names even, not folders in case sensitive systems (for me). I saw differences in console output of the names too I can remember which led me to this conclusion. Maybe it's even MacOS only, not sure.

Anyway, it's certainly not really fixable by GrumPHP, or at least it shouldn't. 

Since I changed the volume name to something without capitals (the first char can be a capital somehow :P) GrumPHP is just working fine :).
